fix(sdk-python): e2e tests don't clean up created prompts on teardown — production license quota hits 10k cap

## Symptom

`sdk-python-ci` has been failing on **every branch (including main)** for 24+ hours with `403 resource_limit_exceeded` errors:

> `License limit of 10000 prompts reached. To increase your limits, upgrade your plan at https://app.langwatch.ai/settings/subscription`

This is hitting the python-sdk test project on `app.langwatch.ai` (auth via `PYTHON_SDK_LANGWATCH_API_KEY` GH secret). All `tests/e2e/test_fetch_policies_e2e.py` and `tests/e2e/test_prompt_tags_e2e.py` tests fail at the **`langwatch.prompts.create(...)`** call before any test assertion runs.

Example from the failing run on PR #3083 (15 failures, all the same shape):
```
FAILED tests/e2e/test_fetch_policies_e2e.py::TestFetchPoliciesE2E::test_materialized_first_falls_back_to_api_when_local_missing
  langwatch.generated.langwatch_rest_api_client.errors.UnexpectedStatus: Unexpected status code: 403
  Response content: {"error":"resource_limit_exceeded","message":"License limit of 10000 prompts reached"...}
```

## Immediate mitigation

A one-shot ops workflow (`.github/workflows/delete-test-prompts.yml` on branch `ops/wipe-test-prompts`) wipes all prompts on the test project as an emergency reset. **This is a bandaid** — the underlying leak will refill the quota over time.

## Root cause

Tests create prompts via `langwatch.prompts.create(...)` but don't reliably clean them up on teardown. Most tests already use a `try/finally` cleanup pattern (good!), but there are gaps:

### Confirmed leaker
**`python-sdk/tests/e2e/test_fetch_policies_e2e.py::test_materialized_only_returns_local_prompt_without_api_call`** (line 275)
- Creates a prompt with `langwatch.prompts.create(...)` 
- Never deletes it (no `try/finally`, no fixture cleanup)
- Leaks exactly **1 prompt per CI run**

```python
# python-sdk/tests/e2e/test_fetch_policies_e2e.py:275 (current — LEAKS)
server_prompt = langwatch.prompts.create(
    handle=handle, prompt="Hello from MATERIALIZED_ONLY policy test"
)
# ... assertions ...
# end of test, no delete  ← LEAK
```

### Tests that *do* clean up properly (reference pattern)
- All 13 tests in `tests/e2e/test_prompt_tags_e2e.py` — `try/finally` per test
- 3 of 4 prompt-creating tests in `test_fetch_policies_e2e.py` (lines 181-204, 226-253, 320-362)

### Additional leak sources to investigate
Even tests that *do* have `try/finally` cleanup will leak in these cases:
1. **`langwatch.prompts.delete()` itself fails** (e.g., transient 5xx) — failures are silently swallowed in some places
2. **CI run cancellation** (e.g., concurrency cancel-in-progress) leaves the create complete but skips finally
3. **Test process killed** (OOM, timeout) — same as above
4. **Multiple PRs / branches** running concurrently — `cancel-in-progress: true` is set in `.github/workflows/sdk-python-ci.yml:21-22`, so cancelled runs are common

10k prompts / 1 leak per run × ~30 runs/day × ~30 days ≈ matches the timeline.

## Suggested fix

**Phase 1 — close the obvious leak:**
- Add `try/finally` cleanup to `test_materialized_only_returns_local_prompt_without_api_call` matching the pattern already used in the same file (lines 180-208):
  ```python
  server_prompt = langwatch.prompts.create(handle=handle, prompt="...")
  try:
      # test body
  finally:
      try:
          langwatch.prompts.delete(server_prompt.id)
      except Exception as e:
          logger.warning("Failed to delete prompt %s: %s", server_prompt.id, e)
  ```

**Phase 2 — make cleanup robust against crashes/cancellations:**
- Replace per-test `try/finally` with a **module-scoped pytest fixture** that tracks created prompts and tears them down even if individual tests crash:
  ```python
  @pytest.fixture
  def prompt_factory():
      created: list[str] = []
      def _create(**kwargs):
          p = langwatch.prompts.create(**kwargs)
          created.append(p.id)
          return p
      yield _create
      for pid in created:
          try:
              langwatch.prompts.delete(pid)
          except Exception:
              pass
  ```
- Also consider a **session-scoped autouse fixture** that runs once at the end and lists/deletes any prompts whose handles match a known test prefix (e.g., `test-*` or `e2e-*`), as a safety net for orphans from cancelled runs.

**Phase 3 — production guardrail:**
- Add a daily scheduled workflow that wipes prompts older than 24h on the python-sdk test project (would prevent quota exhaustion regardless of whether per-test cleanup is bullet-proof).

## Acceptance criteria
- [ ] `test_materialized_only_returns_local_prompt_without_api_call` cleans up its created prompt
- [ ] All e2e tests use a centralized cleanup fixture rather than per-test `try/finally`
- [ ] Prompt cleanup is resilient to test crashes and CI cancellations
- [ ] (Optional) Daily scheduled cleanup workflow exists as a safety net
- [ ] `sdk-python-ci` runs cleanly on main for 7 consecutive days without hitting the quota

## Investigation notes for whoever picks this up
- The temporary emergency-reset workflow lives at `.github/workflows/delete-test-prompts.yml` on branch `ops/wipe-test-prompts`. That branch can stay for now as a recovery tool.
- The test project ID and org ID are in the GH secret `PYTHON_SDK_LANGWATCH_API_KEY` — don't hardcode them.
- Auth header is `X-Auth-Token` (see `langwatch/src/app/api/middleware/auth.ts:19`).

--- a/python-sdk/tests/e2e/conftest.py
+++ b/python-sdk/tests/e2e/conftest.py
@@ -66,6 +66,50 @@ def _delete_prompt_with_narrow_handling(prompt_id: str, context: str) -> None:
         )
 
 
+def _delete_tag_with_narrow_handling(name: str, context: str) -> None:
+    """
+    Delete a single tag, distinguishing 404 (already gone) from real errors.
+
+    ``langwatch.prompts.tags.delete`` routes 404 responses through the same
+    ``unwrap_response`` helper as prompts, which raises
+    ``ValueError("Prompt not found: ...")`` regardless of the resource type
+    (see ``langwatch.prompts.errors``).  Tag teardown routinely encounters
+    this when a test deletes the tag itself as part of the assertion
+    (e.g. ``test_delete_tag_cascades_to_assignments``).
+
+    Never raises — callers iterate lists and must clean up all remaining names.
+    """
+    try:
+        langwatch.prompts.tags.delete(name)
+        logger.info(
+            "%s: deleted tag on teardown",
+            context,
+            extra={"tag_name": name},
+        )
+    except ValueError as exc:
+        if str(exc).startswith("Prompt not found"):
+            logger.debug(
+                "%s: tag already gone (404), skipping",
+                context,
+                extra={"tag_name": name},
+            )
+        else:
+            logger.warning(
+                "%s: failed to delete tag (ValueError: %s), continuing cleanup",
+                context,
+                exc,
+                extra={"tag_name": name},
+            )
+    except Exception as exc:  # noqa: BLE001 - teardown must never raise.
+        logger.warning(
+            "%s: failed to delete tag (%s: %s), continuing cleanup",
+            context,
+            type(exc).__name__,
+            exc,
+            extra={"tag_name": name},
+        )
+
+
 @pytest.fixture(scope="session")
 def _session_prompt_registry() -> List[str]:
     """
@@ -160,7 +204,4 @@ def tag_factory() -> Callable[[str], Any]:
     yield _create
 
     for name in tracked_names:
-        try:
-            langwatch.prompts.tags.delete(name)
-        except Exception as exc:
-            logger.warning("Failed to delete tag %s during teardown: %s", name, exc)
+        _delete_tag_with_narrow_handling(name, "tag_factory")

--- a/python-sdk/tests/e2e/conftest.py
+++ b/python-sdk/tests/e2e/conftest.py
@@ -6,6 +6,8 @@ Provides:
   cleanup in teardown, even when the test body raises (issue #3164).
 - _session_prompt_registry: session-scoped safety-net that deletes any prompts
   created by this session that were not cleaned up by their per-test teardown.
+- tag_factory: function-scoped fixture that creates prompt tags and deletes
+  them on teardown (issue #3108).
 """
 
 import logging
@@ -137,3 +139,28 @@ def prompt_factory(_session_prompt_registry: List[str]) -> Callable[..., Any]:
             _session_prompt_registry.remove(prompt_id)
         except ValueError:
             pass  # already removed or never added (shouldn't happen)
+
+
+@pytest.fixture
+def tag_factory() -> Callable[[str], Any]:
+    """
+    Function-scoped factory fixture for creating prompt tags with guaranteed cleanup.
+
+    Returns a callable that accepts a tag ``name`` and creates it via
+    ``langwatch.prompts.tags.create(name)``.  Every tag created through the
+    factory is deleted during fixture teardown, even when the test body raises.
+    """
+    tracked_names: List[str] = []
+
+    def _create(name: str) -> Any:
+        tag = langwatch.prompts.tags.create(name)
+        tracked_names.append(name)
+        return tag
+
+    yield _create
+
+    for name in tracked_names:
+        try:
+            langwatch.prompts.tags.delete(name)
+        except Exception as exc:
+            logger.warning("Failed to delete tag %s during teardown: %s", name, exc)

--- a/python-sdk/tests/e2e/test_prompt_tags_e2e.py
+++ b/python-sdk/tests/e2e/test_prompt_tags_e2e.py
@@ -347,7 +347,7 @@ class TestPromptTagsE2E:
             except Exception as e:
                 logger.warning("Failed to delete tag %s: %s", tag_b, e)
 
-    def test_delete_tag_cascades_to_assignments(self, prompt_factory):
+    def test_delete_tag_cascades_to_assignments(self, prompt_factory, tag_factory):
         """
         GIVEN a prompt with a custom tag assigned
         WHEN I delete the tag
@@ -357,30 +357,24 @@ class TestPromptTagsE2E:
         tag_name = f"e2e-cascade-tag-{uuid4().hex[:8]}"
         handle = f"e2e-cascade-prompt-{uuid4().hex[:8]}"
 
-        langwatch.prompts.tags.create(tag_name)
+        tag_factory(tag_name)
         created = prompt_factory(handle=handle, prompt="Cascade test")
 
-        try:
-            langwatch.prompts.tags.assign(
-                handle,
-                tag=tag_name,
-                version_id=created.version_id,
-            )
+        langwatch.prompts.tags.assign(
+            handle,
+            tag=tag_name,
+            version_id=created.version_id,
+        )
 
-            langwatch.prompts.tags.delete(tag_name)
+        langwatch.prompts.tags.delete(tag_name)
 
-            tags_after = langwatch.prompts.tags.list()
-            tag_names_after = [t["name"] for t in tags_after]
-            assert tag_name not in tag_names_after
+        tags_after = langwatch.prompts.tags.list()
+        tag_names_after = [t["name"] for t in tags_after]
+        assert tag_name not in tag_names_after
 
-            fetched = langwatch.prompts.get(handle)
-            assert fetched is not None
-            assert fetched.handle == handle
-        finally:
-            try:
-                langwatch.prompts.tags.delete(tag_name)
-            except Exception:
-                pass
+        fetched = langwatch.prompts.get(handle)
+        assert fetched is not None
+        assert fetched.handle == handle
 
     def test_shorthand_syntax_passes_through_as_id(self, prompt_factory):
         """

